ch.vorburger.osgi.gradle/src/main/java/ch/vorburger/osgi/maven/internal/MavenBuildService.java

```diff
@@ -0,0 +1,175 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ch.vorburger.osgi.maven.internal;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Arrays;
+
+import ch.vorburger.osgi.gradle.internal.BuildService;
+import ch.vorburger.osgi.gradle.internal.BuildServiceListener;
+import ch.vorburger.osgi.gradle.internal.concurrent.ExecutorServiceProviderSingleton;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.maven.shared.invoker.DefaultInvocationRequest;
+import org.apache.maven.shared.invoker.DefaultInvoker;
+import org.apache.maven.shared.invoker.InvocationRequest;
+import org.apache.maven.shared.invoker.Invoker;
+import org.apache.maven.shared.invoker.MavenInvocationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * Implementation of BuildService for Gradle.
+ */
+public class MavenBuildService implements BuildService {
+
+    private final ListeningExecutorService executorService;
+    private WatchService watcher;
+
+
+    public MavenBuildService() {
+        this(ExecutorServiceProviderSingleton.INSTANCE.newCachedThreadPool("MavenBuildService"));
+    }
+
+    public MavenBuildService(ListeningExecutorService executorService) {
+        try {
+            this.watcher = FileSystems.getDefault().newWatchService();
+        } catch (IOException ex) {
+            throw new RuntimeException("could not create watch service", ex);
+        }
+        this.executorService = executorService;
+    }
+    
+    private ListenableFuture<Void> build(File projectDirectory, String[] tasks, boolean continuous, BuildServiceListener listener) {
+        return executorService.submit(() -> {
+            build(projectDirectory, tasks, listener);
+            if (continuous) {
+                Path path = projectDirectory.toPath();
```

Comment:
@edewit FYI I built something like this in https://github.com/vorburger/HoTea/tree/master/ch.vorburger.hotea/src/main/java/ch/vorburger/hotea/watchdir - if you would like to copy/paste that utility from there to here and use it here, I'm all for it!